src/csharp/Microsoft.Spark.E2ETest/IpcTests/JvmThreadPoolTests.cs

```diff
@@ -0,0 +1,101 @@
+        private readonly IJvmBridge _jvmBridge;
+
+        public JvmThreadPoolTests(SparkFixture fixture)
+        {
+            _spark = fixture.Spark;
+            _jvmBridge = ((IJvmObjectReferenceProvider)_spark).Reference.Jvm;
+        }
+
+        /// <summary>
+        /// Test that the active SparkSession is thread-specific.
+        /// </summary>
+        [Fact]
+        public void TestThreadLocalSessions()
+        {
+            SparkSession.ClearActiveSession();
+
+            void testChildThread(string appName)
+            {
+                var thread = new Thread(() =>
+                {
+                    Assert.Null(SparkSession.GetActiveSession());
+
+                    SparkSession.SetActiveSession(
+                        SparkSession.Builder().AppName(appName).GetOrCreate());
+
+                    // Since we are in the child thread, GetActiveSession() should return the child
+                    // SparkSession.
+                    var activeSession = SparkSession.GetActiveSession();
+                    Assert.NotNull(activeSession);
+                    Assert.Equal(appName, activeSession.Conf().Get("spark.app.name", null));
+                });
+
+                thread.Start();
+                while (thread.IsAlive)
+                {
+                    Thread.Sleep(1000);
+                }
+            }
+
+            for (var i = 0; i < 5; i++)
+            {
+                testChildThread(i.ToString());
+            }
+
+            Assert.Null(SparkSession.GetActiveSession());
+        }
+
+        /// <summary>
+        /// Add and remove a thread via the JvmThreadPool.
+        /// </summary>
+        [Fact]
+        public void TestAddRemoveThread()
+        {
+            var threadPool = new JvmThreadPool(_jvmBridge, TimeSpan.FromMinutes(30));
+
+            var thread = new Thread(() => _spark.Sql("SELECT TRUE"));
+            thread.Start();
+
+            Assert.True(threadPool.TryAddThread(thread));
+            // Subsequent call should return false, because the thread has already been added.
+            Assert.False(threadPool.TryAddThread(thread));
+
+            thread.Join();
+
+            Assert.True(threadPool.TryRemoveThread(thread.ManagedThreadId));
+
+            // Subsequent call should return false, because the thread has already been removed.
+            Assert.False(threadPool.TryRemoveThread(thread.ManagedThreadId));
+        }
+
+        /// <summary>
+        /// Create a Spark worker thread in the JVM ThreadPool then remove it directly through
+        /// the JvmBridge.
+        /// </summary>
+        [Fact]
+        public void TestThreadRm()
+        {
+            // Create a thread and ensure that it is initialized in the JVM ThreadPool.
+            var thread = new Thread(() => _spark.Sql("SELECT TRUE"));
+            thread.Start();
+            thread.Join();
+            _jvmBridge.CallStaticJavaMethod("DotnetHandler", "rmThread", thread.ManagedThreadId);
```

Comment:
Should we add an `Assert.False()` here to make sure `TryRemoveThread()` fails after calling `rmThread`?

Reply:
The JVM method `rmThread` is a no-op if `rmThread` has already been called for that thread on the JVM side.  So calling `rmThread` manually here doesn't actually influence the return value of `TryRemoveThread()`.  For this test, I just wanted to check that `rmThread` returns successfully without throwing exceptions.